src/main/telemetry/smartport.c

```diff
@@ -278,6 +363,175 @@ void checkSmartPortTelemetryState(void)
+
+        smartPortMspReplyPending = true;
+    }
+    else {
+        //TODO: send ACK reply to avoid re-transmission?
+    }
+}
+
+/**
+ * Frame format:
+ * - Header: 1 byte
+ *   - Version: 3 bits
+ *   - Start-flag: 1 bit
+ *   - CSeq: 4 bits
+ *
+ * - MSP payload:
+ *   - Size: 1 Byte
+ *   - Type: 1 Byte
+ *   - payload...
+ *   - CRC
+ */
+void handleSmartPortMspFrame(uint8_t* frame, uint8_t size)
+{
+    static uint8_t mspBuffer[SMARTPORT_MSP_RX_BUF_SIZE];
+    static uint8_t mspStarted = 0;
+    static uint8_t lastSeq = 0;
+    static uint8_t checksum = 0;
+
+    static mspPacket_t cmd = {
+        .buf = { .ptr = mspBuffer, .end = mspBuffer },
+        .cmd = -1,
+        .result = 0
+    };
+    
+    // re-assemble MSP frame & forward to MSP port when complete
+    uint8_t* p = frame;
+    uint8_t* end = frame + size;
+    
+    uint8_t head = *p++;
+    uint8_t seq = head & SMARTPORT_MSP_SEQ_MASK;
+    uint8_t version = (head & SMARTPORT_MSP_VER_MASK) >> SMARTPORT_MSP_VER_SHIFT;
+
+    if (version != SMARTPORT_MSP_VERSION) {
+        // TODO: should a version mismatch error
+        //       be sent back to the transmitter?
+        resetMspPacket(&cmd);
+        return;
+    }
+
+    // check start-flag
+    if (head & SMARTPORT_MSP_START_FLAG) {
+
+        //TODO: if (p_size > SMARTPORT_MSP_RX_BUF_SIZE) error!
+        uint8_t p_size = *p++;
+        cmd.cmd = *p++;
+        cmd.result = 0;
+
+        cmd.buf.ptr = mspBuffer;
+        cmd.buf.end = mspBuffer + p_size;
+
+        checksum = p_size ^ cmd.cmd;
+        mspStarted = 1;
+    }
+    else if (!mspStarted) {
+        // no start packet yet, throw this one away
+        return;
+    }
+    else if (((lastSeq + 1) & SMARTPORT_MSP_SEQ_MASK) != seq) {
+        // packet loss detected!
+        resetMspPacket(&cmd);
+        return;
+    }
+
+    // copy payload bytes
+    while ((p < end) && sbufBytesRemaining(&cmd.buf)) {
+        checksum ^= *p;
+        sbufWriteU8(&cmd.buf,*p++);
+    }
+
+    // reached end of smart port frame
+    if (p == end) {
+        lastSeq = seq;
+        return;
+    }
+
+    // last byte must be the checksum
+    if (checksum != *p) {
+        resetMspPacket(&cmd);
+        return;
+    }
+
+    // end of MSP packet reached
+    mspStarted = 0;
+    sbufSwitchToReader(&cmd.buf,mspBuffer);
+
+    processMspPacket(&cmd);
+}
+
+bool smartPortSendMspReply()
+{
+    static uint8_t checksum = 0;
+    static uint8_t seq = 0;
+
+    uint8_t packet[SMARTPORT_PAYLOAD_SIZE];
+    uint8_t* p = packet;
+    uint8_t* end = p + SMARTPORT_PAYLOAD_SIZE;
+    
+    sbuf_t* txBuf = &smartPortMspReply.buf;
+
+    // detect first reply packet
+    if (txBuf->ptr == smartPortMspTxBuffer) {
+
+        uint8_t size = sbufBytesRemaining(txBuf);
+
+        // header
+        *p++ = SMARTPORT_MSP_VERSION_S | SMARTPORT_MSP_START_FLAG | (seq++ & SMARTPORT_MSP_SEQ_MASK);
+
+        *p++ = size;
+        checksum = sbufBytesRemaining(txBuf) ^ smartPortMspReply.cmd;        
```

Comment:
By the way, "size" IS checksumed. Just not when it's equal 0. But I should rather use "size" instead of "sbufBytesRemaining(txBuf)" to make that more clear.